File: week2/readability.c
// READABILITY: https://cs50.harvard.edu/x/2023/psets/2/readability/

#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string words);
int count_sentences(string words);

//MAIN
int main(void)
{
    string text = get_string("Text: ");
    float l = (100 * ((float) count_letters(text) / (float) count_words(text)));
    float s = (100 * ((float) count_sentences(text) / (float) count_words(text)));
    //Coleman-Liau index
    float coleman_liau = (0.0588 * l) - (0.296 * s) - 15.8;
    int cl = (round(coleman_liau));

    if (cl >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (cl < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %d\n", cl);
    }
    //printf("%d letters\n", count_letters(text));
    //printf("%d words\n", count_words(text));
    //printf("%d sentences\n", count_sentences(text));
}


//function to count only letters
int count_letters(string text)
{
    int count = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isalpha(text[i]))
        {
            count++;
        }
    }

    return count;
}

//function to count the words
//words like sister-in-law is considered to be one word
int count_words(string words)
{
    int len = strlen(words);
    int count = 0;
    char nonwords[] = " .,;\n\t";

    for (int i = 0; i < len; i++)
    {
        while (i < len)
        {
            if (strchr(nonwords, words[i]) != NULL)
            {
                break;
            }
            i++;
        }

        count++;

        while (i < len)
        {
            if (strchr(nonwords, words[i]) == NULL)
            {
                break;
            }
            i++;
        }
    }
    return count;
}

//function to count the sentences
int count_sentences(string words)
{
    int len = strlen(words);
    int count = 0;
    char nonwords[] = ".!?";

    for (int i = 0; i < len; i++)
    {
        while (i < len)
        {
            if (strchr(nonwords, words[i]) != NULL)
            {
                break;
            }
            i++;
        }

        count++;

        while (i < len)
        {
            if (strchr(nonwords, words[i]) == NULL)
            {
                break;
            }
            i++;
        }
    }
    return count;
}

File: week6/problem_set_6/readability.py
# READABILITY: https://cs50.harvard.edu/x/2023/psets/6/readability/

from cs50 import get_string

text = get_string("Text: ")

# count the number of letters in the text
count_letters = 0
for c in text:
    if c.isalpha():
        count_letters += 1

# count the number of words in the text
count_words = len(text.split())

# count the number of sentences in the text
count_sentences = text.count(".") + text.count("!") + text.count("?")

# the average number of letters per 100 words in the text
L = count_letters / count_words * 100

# the average number of sentences per 100 words in the text
S = count_sentences / count_words * 100

# Coleman-Liau index
coleman_liau = 0.0588 * L - 0.296 * S - 15.8

# print the result
if coleman_liau >= 16:
    print("Grade 16+")
elif coleman_liau < 1:
    print("Before Grade 1")
else:
    print(f"Grade {round(coleman_liau)}")
